src/serac/numerics/functional/boundary_integral_kernels.hpp

```diff
@@ -257,22 +317,43 @@ void action_of_gradient_kernel(const double* dU, double* dR, derivatives_type* q
 
   // mfem provides this information in 1D arrays, so we reshape it
   // into strided multidimensional arrays before using
-  constexpr int                                   nqp = num_quadrature_points(geom, Q);
-  auto                                            du  = reinterpret_cast<const typename trial_element::dof_type*>(dU);
-  auto                                            dr  = reinterpret_cast<typename test_element::dof_type*>(dR);
-  static constexpr TensorProductQuadratureRule<Q> rule{};
+  constexpr int                            nqp = num_quadrature_points(geom, Q);
+  auto                                     du  = reinterpret_cast<const typename trial_element::dof_type*>(dU);
+  auto                                     dr  = reinterpret_cast<typename test_element::dof_type*>(dR);
+  constexpr TensorProductQuadratureRule<Q> rule{};
 
-  // for each element in the domain
-  for (uint32_t e = 0; e < num_elements; e++) {
-    // (batch) interpolate each quadrature point's value
-    auto qf_inputs = trial_element::interpolate(du[elements[e]], rule);
+  auto e_range = RAJA::TypedRangeSegment<std::size_t>(0, num_elements);
 
-    // (batch) evalute the q-function at each quadrature point
-    auto qf_outputs = batch_apply_chain_rule(qf_derivatives + e * nqp, qf_inputs);
+  using qf_inputs_type = decltype(trial_element::template interpolate_output_helper<Q>());
 
-    // (batch) integrate the material response against the test-space basis functions
-    test_element::integrate(qf_outputs, rule, &dr[elements[e]]);
-  }
+#ifdef SERAC_USE_CUDA_KERNEL_EVALUATION
+  std::string device_name = "DEVICE";
+#else
+  std::string                      device_name = "HOST";
+#endif
+
+  auto&           rm        = umpire::ResourceManager::getInstance();
+  auto            allocator = rm.getAllocator(device_name);
+  qf_inputs_type* qf_inputs = static_cast<qf_inputs_type*>(allocator.allocate(sizeof(qf_inputs_type) * num_elements));
```

Comment:
Can this allocate/deallocate be used with `std::unique_ptr<qf_inputs>`?

Maybe you could define:
```c++
template<typename MyType>
::std::unique_ptr<MyType>
allocateOnHardware(::std::size_t size) {
// #ifdef to get device name
// get Umpire instance
// get appropriate allocator
   return ::std::unique_ptr<MyType[]>(static_cast<MyType*>(allocator.allocate(size), [&](void *ptr) { rm.deallocate(ptr); });
}
```